### ISSUE TITLE
No speedup using XGBClassifier with GPU support

### ISSUE THREAD

**Issue by h-amirkhani** [dmlc/xgboost#2819]
In the following code, I try to search over different hyper-parameters of xgboost.

    param_test1 = {
     'max_depth':list(range(3,10,2)),
     'min_child_weight':list(range(1,6,2))
    }
    predictors = [x for x in train_data.columns if x not in ['target', 'id']]
    gsearch1 = GridSearchCV(estimator=XGBClassifier(learning_rate =0.1, n_estimators=100, max_depth=5,
                                                    min_child_weight=1, gamma=0, subsample=0.8, colsample_bytree=0.8,
                                                    objective= 'binary:logistic', n_jobs=4, scale_pos_weight=1, seed=27, 
                                                    kvargs={'tree_method':'gpu_hist'}),
                        param_grid=param_test1, scoring='roc_auc', n_jobs=4, iid=False, cv=5, verbose=2)
    gsearch1.fit(train_data[predictors], train_data['target'])

Even though I use `kvargs={tree_method':'gpu_hist'}`, I get no speedup in the implementation. According to the `nvidia-smi`, the GPU is not much involved in the computation:

    +-----------------------------------------------------------------------------+
    | NVIDIA-SMI 375.66                 Driver Version: 375.66                    |
    |-------------------------------+----------------------+----------------------+
    | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
    | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
    |===============================+======================+======================|
    |   0  GeForce GTX 1080    Off  | 0000:01:00.0      On |                  N/A |
    |  0%   39C    P8    10W / 200W |    338MiB /  8112MiB |      1%      Default |
    +-------------------------------+----------------------+----------------------+
                                                                                   
    +-----------------------------------------------------------------------------+
    | Processes:                                                       GPU Memory |
    |  GPU       PID  Type  Process name                               Usage      |
    |=============================================================================|
    |    0       961    G   /usr/lib/xorg/Xorg                             210MiB |
    |    0      1675    G   compiz                                         124MiB |
    |    0      2359    G   /usr/lib/firefox/firefox                         2MiB |
    +-----------------------------------------------------------------------------+
I have installed the GPU supported xgboost using the following commands in Ubuntu:

    $ git clone --recursive https://github.com/dmlc/xgboost
    $ mkdir build
    $ cd build
    $ cmake .. -DUSE_CUDA=ON
    $ make -j

What is the possible reason?

**Comment by RAMitchell** [comment]
If you look at the XGBoost Python API documentation you will see that arguments passed via kwargs are not guaranteed to work with sklearn functions. 

This is currently expected behaviour. We could consider adding the tree method parameter directly to the XGBClassifier to allow this future.

**Comment by h-amirkhani** [comment]
Thanks for your comment.

**Comment by up201007037** [comment]
It does work. I tested that, and you can get gpu support with sklearn. You need to pass the parameter as a regular parameter:

`params_dist = {  
    "n_estimators": st.randint(3, 1000),
    'tree_method':['gpu_hist'], 
    'n_jobs':[-1],
    "max_depth": st.randint(3, 10),
    "learning_rate": np.logspace(-4, -0.3, num=100),
    "colsample_bytree": one_to_left,
    "subsample": one_to_left,
    "gamma": st.uniform(0, 10),
    'reg_alpha': from_zero_positive,
    "min_child_weight": from_zero_positive,
    'predictor':['gpu_predictor']
}`

**Comment by h-amirkhani** [comment]
Thanks. I'll test it again.

**Comment by up201007037** [comment]
Ok, tell me later please if it works.
However, I do not understand why it does not use the gpu so much. I can only use 200/300MB of gpu memory. But still it is 3x faster than the cpu.

P.s do not use gridsearchcv or randomizedcv from sklearn with xgboost. I takes so long to run. Maybe it is a bug. I implemented my own randomizedcv function and it runs 10x faster...

**Comment by h-amirkhani** [comment]
Yes, it works. I ran an experiment and got exactly what you said:
`param = {'max_depth':5, 'objective':'binary:logistic', 'subsample':0.8, 
         'colsample_bytree':0.8, 'eta':0.5, 'min_child_weight':1,
         'tree_method':'gpu_hist'}
num_round = 100

dtrain = xgb.DMatrix(train_X, train_y)
tic = time.time()
model = xgb.train(param, dtrain, num_round)
print('passed time with xgb (gpu): %.3fs'%(time.time()-tic))

xgb_param = {'max_depth':5, 'objective':'binary:logistic', 'subsample':0.8, 
         'colsample_bytree':0.8, 'learning_rate':0.5, 'min_child_weight':1,
         'tree_method':'gpu_hist'}
model = XGBClassifier(**xgb_param)
tic = time.time()
model.fit(train_X, train_y)
print('passed time with XGBClassifier (gpu): %.3fs'%(time.time()-tic))

param = {'max_depth':5, 'objective':'binary:logistic', 'subsample':0.8, 
         'colsample_bytree':0.8, 'eta':0.5, 'min_child_weight':1,
         'tree_method':'hist'}
num_round = 100

dtrain = xgb.DMatrix(train_X, train_y)
tic = time.time()
model = xgb.train(param, dtrain, num_round)
print('passed time with xgb (cpu): %.3fs'%(time.time()-tic))

xgb_param = {'max_depth':5, 'objective':'binary:logistic', 'subsample':0.8, 
         'colsample_bytree':0.8, 'learning_rate':0.5, 'min_child_weight':1,
         'tree_method':'hist'}
model = XGBClassifier(**xgb_param)
tic = time.time()
model.fit(train_X, train_y)
print('passed time with XGBClassifier (cpu): %.3fs'%(time.time()-tic))`

and these are the results:
`passed time with xgb (gpu): 14.420s
passed time with XGBClassifier (gpu): 15.129s
passed time with xgb (cpu): 38.372s
passed time with XGBClassifier (cpu): 38.837s`

About GPU memory usage, I also noticed that XGBClassifier uses only around 300MB of the memory, but xgb can use upto 500MB (still not a complete usage of the memory).

**Comment by wubizhi** [comment]
just not set the paramter "n_jobs",and i will be work,for example
*******************************************************************************
param_depth_child_weight={
    'max_depth':np.arange(5,8,2),
    'min_child_weight':np.arange(1,4,2)
}
params = {
    "learning_rate":0.1,
    "n_setimators":140,
    "max_depth":5,
    "min_child_weight":1,
    "gamma":0,
    "subsample":0.8,
    "colsample_bytree":0.8,
    "objective":'binary:logistic',
    "nthread":4,
    "scale_pos_weight":80,
    "tree_method":'gpu_hist', 
    "predictor":'gpu_predictor',
    "seed":27
}
xgb_depth_child_weight = XGBClassifier(**params)
xgb_search_depth_child_weight = GridSearchCV(estimator=xgb_depth_child_weight,param_grid=param_depth_child_weight,
                                             scoring='roc_auc',iid=False,cv=5,verbose=2)
*******************************************************************************

if you set the pamater nthread in XGBClassifer,you do not set the n_jobs in GridSearchCV.
